File: Data Structures/Arrays/Minimum Swaps 2/minSwaps.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Solution {

    // Complete the minimumSwaps function below.
    static int minimumSwaps(int[] arr) {
        int counter = 0;
        int[] ph = new int[arr.Length];

        for (int i = 0; i < arr.Length; i++) {
            int val = arr[i] - 1;
            ph[val] = i;
        }

        for (int i = 0; i < arr.Length; i++) {
            if (arr[i] != i+1) {
                int l = ph[i];
                int temp = arr[i];
                arr[i] = i+1;
                arr[l] = temp;
                ph[i] = i;
                int val = arr[l] - 1;
                ph[val] = l;
                counter += 1;
            }
            
        }

        return counter;

    }

    static void Main(string[] args) {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        int n = Convert.ToInt32(Console.ReadLine());

        int[] arr = Array.ConvertAll(Console.ReadLine().Split(' '), arrTemp => Convert.ToInt32(arrTemp))
        ;
        int res = minimumSwaps(arr);

        textWriter.WriteLine(res);

        textWriter.Flush();
        textWriter.Close();
    }
}

File: Data Structures/Arrays/Minimum Swaps 2/minSwaps.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the minimumSwaps function below.
function minimumSwaps(arr) {
    var counter = 0;
    var ph = [];

    for (let i = 0; i < arr.length; i++) {
        var val = arr[i] - 1;
        ph[val] = i;
    }

    for (let i = 0; i < arr.length; i++) {
        if (arr[i] != i + 1) {
            var l = ph[i];
            var temp = arr[i];
            arr[i] = i+1;
            arr[l] = temp;
            ph[i] = i;
            var val = arr[l] - 1;
            ph[val] = l;
            counter += 1;
        }
    }
    return counter;


}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine(), 10);

    const arr = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));

    const res = minimumSwaps(arr);

    ws.write(res + '\n');

    ws.end();
}
